### PR TITLE
feat(wallet): implement method to refresh the wallet session

### LINKED ISSUE #1927: Allow refreshing session before it expires

### PR DESCRIPTION
**Pull request by gabaldon**
close #1927

### CODE REVIEW

File: wallet/src/actors/app/handlers/unlock_wallet.rs [hunk 1]
```diff
@@ -31,7 +31,7 @@ impl Handler<UnlockWalletRequest> for app::App {
     fn handle(&mut self, msg: UnlockWalletRequest, _ctx: &mut Self::Context) -> Self::Result {
         let f = self.unlock_wallet(msg.wallet_id, msg.password).map_ok(
             |types::UnlockedWallet { data, session_id }, slf, ctx| {
-                slf.set_session_to_expire(session_id.clone()).spawn(ctx);
+                slf.set_session_to_expire(session_id.clone()).unwrap().spawn(ctx);
```
Comment (mariocao) [inline wallet/src/actors/app/handlers/unlock_wallet.rs:34]:
```suggestion
                slf.set_session_to_expire(session_id.clone()).expect("Session id should exist after wallet unlock").spawn(ctx);
```

File: wallet/src/actors/app/handlers/refresh_session.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+impl Message for RefreshSessionRequest {
+    type Result = Result<RefreshSessionResponse, app::Error>;
+}
+
+impl Handler<RefreshSessionRequest> for app::App {
+    type Result = <RefreshSessionRequest as Message>::Result;
+
+    fn handle(&mut self, msg: RefreshSessionRequest, ctx: &mut Self::Context) -> Self::Result {
+      let mut session = self.state.sessions.get_mut(&msg.session_id).unwrap();
```
Comment (mariocao) [inline wallet/src/actors/app/handlers/refresh_session.rs:25]: Avoid this `unwrap` by considering the case that the `get_mut` returns a `None`, in which case the `handle` could return a `false`. :)

File: wallet/src/actors/app/methods.rs [hunk 3]
```diff
@@ -447,20 +447,37 @@ impl App {
...
-        TimerFunc::new(
+        Ok(TimerFunc::new(
             self.params.session_expires_in,
-            move |slf: &mut Self, _ctx| match slf.close_session(session_id.clone()) {
-                Ok(_) => log::info!("Session {} closed", session_id),
-                Err(err) => log::error!("Session {} couldn't be closed: {}", session_id, err),
-            },
-        )
+            move |slf: &mut Self, _ctx| {
+                let mut session = slf.state.sessions.get_mut(&session_id).unwrap();
```
Comment (mariocao) [inline wallet/src/actors/app/methods.rs:469]: Avoid this unwrap, if `None` this function shoudl return an `Error`.

**Review by clbartoli — APPROVED** [review]

File: wallet/src/actors/app/handlers/refresh_session.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        let mut session = self
+            .state
+            .sessions
+            .get_mut(&msg.session_id)
+            .ok_or(app::Error::SessionNotFound)?;
+
+        if !session.session_extended {
+            session.session_extended = true;
+            self.set_session_to_expire(msg.session_id.clone())
+                .unwrap()
```
Comment (mariocao) [inline wallet/src/actors/app/handlers/refresh_session.rs:33]:
```suggestion
                .expect("Session id should exist")
```

File: wallet/src/actors/app/handlers/refresh_session.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                .spawn(ctx);
+
+            Ok(RefreshSessionResponse { success: true })
+        } else {
+            Ok(RefreshSessionResponse { success: false })
+        }
```
Comment (mariocao) [inline wallet/src/actors/app/handlers/refresh_session.rs:39]:
```suggestion
        let mut session = self
            .state
            .sessions
            .get_mut(&msg.session_id)
            .ok_or(app::Error::SessionNotFound)?;

        if !session.session_extended {
            session.session_extended = true;
            self.set_session_to_expire(msg.session_id.clone())?.spawn(ctx);

            Ok(RefreshSessionResponse { success: true })
        } else {
            Ok(RefreshSessionResponse { success: false })
        }
```

File: wallet/src/actors/app/handlers/refresh_session.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            .state
+            .sessions
+            .get_mut(&msg.session_id)
+            .ok_or(app::Error::SessionNotFound)?;
+
+        if !session.session_extended {
+            session.session_extended = true;
+            self.set_session_to_expire(msg.session_id.clone())
+                .unwrap()
+                .spawn(ctx);
```
Comment (mariocao) [inline wallet/src/actors/app/handlers/refresh_session.rs:34]:
```suggestion
            self.set_session_to_expire(msg.session_id.clone())?.spawn(ctx);
```

**Review by mariocao — APPROVED** [review]

**Pull request merged**